spec/ruby/optional/capi/ext/io_spec.c

```diff
@@ -168,13 +180,23 @@ VALUE io_spec_rb_thread_wait_fd(VALUE self, VALUE io) {
 }
 
 VALUE io_spec_rb_wait_for_single_fd(VALUE self, VALUE io, VALUE events, VALUE secs, VALUE usecs) {
-  int fd = io_spec_get_fd(io);
+#ifdef RUBY_VERSION_IS_3_0
+  VALUE timeout = Qnil;
+  if (!NIL_P(secs)) {
+      timeout = rb_float_new((double)FIX2INT(secs) + (0.000001f * FIX2INT(usecs)));
+  }
+  VALUE result = rb_io_wait(io, events, timeout);
+  if (result == Qfalse) return INT2FIX(0);
+  else return result;
+#else
```

Comment:
This is not ideal, because it means we won't test `rb_wait_for_single_fd` anymore on Ruby >= 3.
So it'd be better to define a separate function/method and test that separately.

For the other changes above it seems fine because they are all trivial renames or helpers, but it's not that straightforward here.

Reply:
`rb_wait_for_single_fd` should be deprecated and removed. How should we handle it?

Reply:
Every time you call it, the fiber scheduler has to make a temporary IO object.

Reply:
I think deprecating this would need feedback from other core devs or matz.
Practically, you'd use one of the `DEPRECATED` macros, and the specs would stop testing that method when it's removed.

Reply:
Also this PR title doesn't seem really connected to deprecating functions

Reply:
The scope of this PR is pretty big and I'll do my best to narrow it down, maybe introduce the deprecations in a separate PR. However, there are a ton of things that need to happen simultaneously in order to do this work because I don't think then justification for doing them separately would be easy to make a case/argument.

Basically, the entire `struct rb_io` is a public interface, and this PR tries to hide all the details. Whether we can accept it I don't know, but we can't make any forward progress while it remains that way.

Reply:
Agreed it'd be good to make `struct rb_io` internal. I think the PR title just needs to reflect that if that's the main overarching change.

Reply:
For me, making it internal is a means to an end, which is fixing `IO#close` behaviour/semantics/performance. If I have to explain to Matz, which is easier?

Reply:
I think it needs to be clear to everyone that both `IO#close` and `struct rb_io is made internal` are done by this PR/Redmine ticket (`struct rb_io is made internal` is not an implementation detail, it will affect gems).

Reply:
> struct rb_io is made internal is not an implementation detail change, it will affect gems.

Do you know any examples of this?

Reply:
Usages of `RB_IO_POINTER` or `GetOpenFile` seem likely problematic, if they want to access any field of the returned `rb_io_t*`.
`GetOpenFile` is used in a bunch of C extension gems.

Reply:
```
$ gem-codesearch 'GetOpenFile\b' | wc -l
974
```
Not all might access fields of it, but I guess many do.

Reply:
We can add deprecation warnings and alternatives. Most of them are probably just accessing the descriptor.